### PR TITLE
Finalize AST for v3.0 (part 2)

### PR DESCRIPTION
**Pull request by adams85**
While working on the AST to Javascript feature, I identified a few further problems with AST. Hopefully, this PR sorts them out once and for all.

Besides several minor inconsistencies/anomalies, there are some conceptual issues with the AST class hierarchy.

Before the PR it looks like this:

```
Node
├─ClassBody
├─ClassElement
│ ├─ClassProperty
│ │ ├─MethodDefinition
│ │ └─PropertyDefinition
│ └─StaticBlock
├─Decorator
├─ImportAttribute
├─StatementListItem
│ ├─Expression
│ │ ├─ArrayExpression
│ │ ├─ArrowFunctionExpression
│ │ ├─ArrowParameterPlaceHolder
│ │ ├─AssignmentExpression
│ │ ├─AssignmentPattern
│ │ ├─AwaitExpression
│ │ ├─BinaryExpression
│ │ ├─BindingPattern
│ │ │ ├─ArrayPattern
│ │ │ └─ObjectPattern
│ │ ├─CallExpression
│ │ ├─ChainExpression
│ │ ├─ClassExpression
│ │ ├─ConditionalExpression
│ │ ├─FunctionExpression
│ │ ├─Identifier
│ │ ├─Import
│ │ ├─Literal
│ │ ├─MemberExpression
│ │ │ ├─ComputedMemberExpression
│ │ │ └─StaticMemberExpression
│ │ ├─MetaProperty
│ │ ├─NewExpression
│ │ ├─ObjectExpression
│ │ ├─PrivateIdentifier
│ │ ├─Property
│ │ ├─RestElement
│ │ ├─SequenceExpression
│ │ ├─SpreadElement
│ │ ├─Super
│ │ ├─TaggedTemplateExpression
│ │ ├─TemplateLiteral
│ │ ├─ThisExpression
│ │ ├─UnaryExpression
│ │ │ └─UpdateExpression
│ │ ├─YieldExpression
│ │ └─JsxExpression
│ │    ├─JsxAttribute
│ │    ├─JsxClosingElement
│ │    ├─JsxClosingFragment
│ │    ├─JsxElement
│ │    ├─JsxEmptyExpression
│ │    ├─JsxExpressionContainer
│ │    ├─JsxIdentifier
│ │    ├─JsxMemberExpression
│ │    ├─JsxNamespacedName
│ │    ├─JsxOpeningElement
│ │    ├─JsxOpeningFragment
│ │    ├─JsxSpreadAttribute
│ │    └─JsxText
│ └─Statement
│    ├─BlockStatement
│    ├─BreakStatement
│    ├─CatchClause
│    ├─ContinueStatement
│    ├─DebuggerStatement
│    ├─Declaration
│    │ ├─ClassDeclaration
│    │ ├─ExportDeclaration
│    │ │ ├─ExportAllDeclaration
│    │ │ ├─ExportDefaultDeclaration
│    │ │ └─ExportNamedDeclaration
│    │ ├─FunctionDeclaration
│    │ ├─ImportDeclaration
│    │ ├─ImportDeclarationSpecifier
│    │ │ ├─ImportDefaultSpecifier
│    │ │ ├─ImportNamespaceSpecifier
│    │ │ └─ImportSpecifier
│    │ └─VariableDeclaration
│    ├─DoWhileStatement
│    ├─EmptyStatement
│    ├─ExportSpecifier
│    ├─ExpressionStatement
│    │ └─Directive
│    ├─ForInStatement
│    ├─ForOfStatement
│    ├─ForStatement
│    ├─IfStatement
│    ├─LabeledStatement
│    ├─Program
│    │ ├─Module
│    │ └─Script
│    ├─ReturnStatement
│    ├─SwitchStatement
│    ├─ThrowStatement
│    ├─TryStatement
│    ├─WhileStatement
│    └─WithStatement
├─SwitchCase
├─TemplateElement
└─VariableDeclarator
```

After the PR we would have:

```
Node
├─AssignmentPattern
├─BindingPattern
│ ├─ArrayPattern
│ └─ObjectPattern
├─CatchClause
├─ClassBody
├─ClassElement
│ ├─ClassProperty
│ │ ├─MethodDefinition
│ │ └─PropertyDefinition
│ └─StaticBlock
├─Decorator
├─ExportSpecifier
├─ImportAttribute
├─ImportDeclarationSpecifier
│ ├─ImportDefaultSpecifier
│ ├─ImportNamespaceSpecifier
│ └─ImportSpecifier
├─Program
│ ├─Module
│ └─Script
├─Property
├─RestElement
├─StatementListItem
│ ├─Expression
│ │ ├─ArrayExpression
│ │ ├─ArrowFunctionExpression
│ │ ├─AssignmentExpression
│ │ ├─AwaitExpression
│ │ ├─BinaryExpression
│ │ │ └─LogicalExpression
│ │ ├─CallExpression
│ │ ├─ChainExpression
│ │ ├─ClassExpression
│ │ ├─ConditionalExpression
│ │ ├─FunctionExpression
│ │ ├─Identifier
│ │ ├─Import
│ │ ├─Literal
│ │ ├─MemberExpression
│ │ │ ├─ComputedMemberExpression
│ │ │ └─StaticMemberExpression
│ │ ├─MetaProperty
│ │ ├─NewExpression
│ │ ├─ObjectExpression
│ │ ├─PrivateIdentifier
│ │ ├─SequenceExpression
│ │ ├─SpreadElement
│ │ ├─Super
│ │ ├─TaggedTemplateExpression
│ │ ├─TemplateLiteral
│ │ ├─ThisExpression
│ │ ├─UnaryExpression
│ │ │ └─UpdateExpression
│ │ ├─YieldExpression
│ │ └─JsxExpression
│ │    ├─JsxAttribute
│ │    ├─JsxClosingElement
│ │    ├─JsxClosingFragment
│ │    ├─JsxElement
│ │    ├─JsxEmptyExpression
│ │    ├─JsxExpressionContainer
│ │    ├─JsxIdentifier
│ │    ├─JsxMemberExpression
│ │    ├─JsxNamespacedName
│ │    ├─JsxOpeningElement
│ │    ├─JsxOpeningFragment
│ │    ├─JsxSpreadAttribute
│ │    └─JsxText
│ └─Statement
│    ├─BlockStatement
│    ├─BreakStatement
│    ├─ContinueStatement
│    ├─DebuggerStatement
│    ├─Declaration
│    │ ├─ClassDeclaration
│    │ ├─ExportDeclaration
│    │ │ ├─ExportAllDeclaration
│    │ │ ├─ExportDefaultDeclaration
│    │ │ └─ExportNamedDeclaration
│    │ ├─FunctionDeclaration
│    │ ├─ImportDeclaration
│    │ └─VariableDeclaration
│    ├─DoWhileStatement
│    ├─EmptyStatement
│    ├─ExpressionStatement
│    │ └─Directive
│    ├─ForInStatement
│    ├─ForOfStatement
│    ├─ForStatement
│    ├─IfStatement
│    ├─LabeledStatement
│    ├─ReturnStatement
│    ├─SwitchStatement
│    ├─ThrowStatement
│    ├─TryStatement
│    ├─WhileStatement
│    └─WithStatement
├─SwitchCase
├─TemplateElement
└─VariableDeclarator
```

@lahma Could you check that these changes are ok from the Jint point of view?

### CODE REVIEW

**Comment by lahma** [comment]
The structure looks better than before and [I was able to make Jint work with it](https://github.com/sebastienros/jint/compare/main...lahma:jint:esprima-next?expand=1). One note at least was that `AssignmentExpression.Left` can also be `MemberExpression`.

File: src/Esprima/Ast/LogicalExpression.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Esprima.Ast
```
Comment (lahma) [inline src/Esprima/Ast/LogicalExpression.cs:1]: nit: could use file-scoped namespace directly for new code

**Review by lahma — APPROVED** [review]

**Comment by lahma** [comment]
Ready to merge when you say so.

File: src/Esprima/Ast/LogicalExpression.cs (hunk 1 above)
Comment (adams85) [reply src/Esprima/Ast/LogicalExpression.cs:1]: Currently namespace style is completely mixed in the codebase. Let's address this with another PR (there's an .editorconfig switch for this).

**Comment by adams85** [comment]
It's ok.

**Comment by adams85** [comment]
Oh wait.  I missed this

> One note at least was that AssignmentExpression.Left can also be MemberExpression.

I need to think this through.

File: src/Esprima/Ast/LogicalExpression.cs (hunk 1 above)
Comment (lahma) [reply src/Esprima/Ast/LogicalExpression.cs:1]: The problem with existing code is that if changed, every line will be changed in files and history (blame) will be hard...

**Comment by adams85** [comment]
Ok, this issue with `AssignmentExpression.Left` was rather theoretic but I fixed it in `AstToJavascriptConverter`. It should be handled this way to be correct conceptually.

Now we can proceed with merging.

**Comment by lahma** [comment]
Awesome work, thank you!

**Pull request merged**